xwiki-commons-core/xwiki-commons-http/src/main/java/org/xwiki/http/internal/XWikiHTTPClient.java

```diff
@@ -0,0 +1,263 @@
+     * the connection manager automatically in all cases relieving individual {@link HttpClientResponseHandler}s from
+     * having to manage resource deallocation internally.
+     *
+     * @param <T> the type of the result determined by the response handler
+     * @param request the request to execute
+     * @param responseHandler the response handler
+     * @return the response object as generated by the response handler.
+     * @throws IOException in case of a problem or the connection was aborted
+     * @throws ClientProtocolException in case of an http protocol error
+     */
+    public <T> T execute(ClassicHttpRequest request, XWikiHTTPClientResponseHandler<? extends T> responseHandler)
+        throws IOException
+    {
+        return execute(request, null, responseHandler);
+    }
+
+    /**
+     * Executes a request using the default context and processes the response using the given response handler. The
+     * content entity associated with the response is fully consumed and the underlying connection is released back to
+     * the connection manager automatically in all cases relieving individual {@link HttpClientResponseHandler}s from
+     * having to manage resource deallocation internally.
+     *
+     * @param <T> the type of the result determined by the response handler
+     * @param request the request to execute
+     * @param credentials the credentials to use for the request
+     * @param responseHandler the response handler
+     * @return the response object as generated by the response handler.
+     * @throws IOException in case of a problem or the connection was aborted
+     * @throws ClientProtocolException in case of an http protocol error
+     */
+    public <T> T execute(ClassicHttpRequest request, XWikiCredentials credentials,
+        XWikiHTTPClientResponseHandler<? extends T> responseHandler) throws IOException
+    {
+        HttpClientContext context = getHttpClientContext(request, credentials);
+
+        return this.client.execute(request, context, response -> responseHandler.handleResponse(response, context));
+    }
+
+    /**
+     * @param request the request to get the context for
+     * @param credentials the credentials to use for the request
+     * @return the context to use for the request
+     * @throws IOException when failing to get the URI of the request
+     */
+    public HttpClientContext getHttpClientContext(HttpRequest request, XWikiCredentials credentials) throws IOException
+    {
+        UsernamePasswordCredentials finalCredentials;
+        if (credentials == null) {
+            if (this.defaultCredentials != null) {
+                finalCredentials = new UsernamePasswordCredentials(this.defaultCredentials.getUserName(),
+                    this.defaultCredentials.getPassword().toCharArray());
+            } else {
+                return null;
```

Comment:
I don't understand this return? If there's no default credentials and no given credentials, the request could still be performed no?

Reply:
> the request could still be performed no?

Definitely, but you are in `#getHttpClientContext` here, not `#execute`. The credential is the only reason why we configure a context right now.

Adding a comment since I guess others might be confused.